File: solutions/dart/binary-search/1/lib/binary_search.dart
import 'package:binary_search/value_not_found_exception.dart';

class BinarySearch {
  List<int> num_list;
  BinarySearch(List<int> this.num_list);

  int find(int number) {
    int end = this.num_list.length - 1;
    int start = 0;
    while (start <= end) {
      int mid = (start + end) ~/ 2;

      if (num_list[mid] == number) {
        // Found it
        return mid;
      } else if (num_list[mid] < number) {
        start = mid + 1;
      } else {
        end = mid - 1;
      }
    }

    throw ValueNotFoundException('value not found');
  }
}

File: solutions/python/binary-search/1/binary_search.py
def find(search_list, value):
    end=len(search_list)
    if end==0:
        raise ValueError('value not in array')    
    
    start=0
    mid=int(end/2)

    while start<=end:
        if mid>=len(search_list):
            raise ValueError('value not in array')
        
        mid_value=search_list[mid]
        if  mid_value< value:
            start=mid+1
            mid=int((start+end)/2)
        elif mid_value > value: 
            end=mid-1
            mid=int((start+end)/2)
        else: # Found!
            return mid
        
    raise ValueError('value not in array')

        
    

